### PR TITLE
Add tool to generate metrics documentation

### PR DESCRIPTION
**Pull request by machadovilaca**
This PR creates a tool to automatically generate a documentation file with HCO metrics information

Example:

```markdown
# Hyperconverged Cluster Operator metrics
This document aims to help users that are not familiar with metrics exposed by the Hyperconverged Cluster Operator.
All metrics documented here are auto-generated by the utility tool `tools/metricsdocs` and reflects exactly what is being exposed.

## Hyperconverged Cluster Operator Metrics List
### kubevirt_hco_out_of_band_modifications_count
Count of out-of-band modifications overwritten by HCO
### kubevirt_hco_unsafe_modification_count
Count of unsafe modifications in the HyperConverged annotations
## Developing new metrics
After developing new metrics or changing old ones, please run `make generate-doc` to regenerate this document.

If you feel that the new metric doesn't follow these rules, please change `tools/metricsdocs` with your needs.
```

Signed-off-by: João Vilaça <jvilaca@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -205,6 +205,9 @@ deploy_cr:
 validate-no-offensive-lang:
 	./hack/validate-no-offensive-lang.sh
 
+metricsdocs:
```
Comment (erkanerol) [inline Makefile:208]: What about adding this into `generate-doc` or adding this into `generate-doc`'s dependent? 

`generate-doc` is triggered by `make sanity` and the sanity checks ensure that the generated doc is up-to-date according to the code. Could you please check this part?

File: docs/metrics.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+This document aims to help users that are not familiar with all metrics exposed by different KubeVirt components.
+All metrics documented here are auto-generated by the utility tool `tools/metricsdocs` and reflects exactly what is being exposed.
+
+## Hyperconverged Cluster Operator Metrics List
+### kubevirt_hco_out_of_band_modifications_count
+Count of out-of-band modifications overwritten by HCO
+### kubevirt_hco_unsafe_modification_count
+Count of unsafe modifications in the HyperConverged annotations
+## Developing new metrics
+After developing new metrics or changing old ones, please run `make metricsdocs` to regenerate this document.
```
Comment (erkanerol) [inline docs/metrics.md:11]: needs to be updated

Comment (machadovilaca) [reply docs/metrics.md]: changed

File: pkg/metrics/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"log"
+	"net/http"
+)
+
+func Handler(MaxRequestsInFlight int) http.Handler {
```
Comment (erkanerol) [inline pkg/metrics/prometheus.go:10]: I am not sure about name of this file. `prometheus.go` doesn't seem so meaningful to me. `handler.go` may be better but I don't have a strong opinion about that.

**Review by erkanerol — COMMENTED** [review]
except for a few little things, looks good to me. @sradco  could you please check this?

File: docs/metrics.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Hyperconverged Cluster Operator metrics
+This document aims to help users that are not familiar with all metrics exposed by different KubeVirt components.
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply docs/metrics.md:2]: Followed the same struct as in https://github.com/kubevirt/kubevirt/blob/07d6dbd1b5ebe54fd611a2a77feb91cf486816c8/tools/doc-generator/doc-generator.go#L25. But I agree with you, the statement is not properly wrong as is but does not reflect the content of the document.

Do I change it to: "This document aims to help users that are not familiar with metrics exposed by the HCO component."?

File: pkg/metrics/prometheus.go (hunk 3 above)
Comment (machadovilaca) [reply pkg/metrics/prometheus.go:10]: The name could also have something related to "scraper". "metricsScraper" or "prometheusScraper" maybe

File: docs/metrics.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (erkanerol) [reply docs/metrics.md:2]: kubevirt/kubevirt contains codes of multiple component like virt-handler, virt-api, virt-operator, virt-launcher etc. `components` refers them. This repo contains code of hco-operator and hco-webhook. We only have metrics in operator side. Therefore, I would use `Hyperconverged Cluster Operator`

File: pkg/metrics/prometheus.go (hunk 3 above)
Comment (erkanerol) [reply pkg/metrics/prometheus.go:10]: I don't think so. This part is not related to scraping. This part is related to serving metrics. It is not a big deal. You can ignore this comment if everything else is OK : )

File: docs/metrics.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (machadovilaca) [reply docs/metrics.md:2]: oh, makes sense, updated

File: tools/metricsdocs/metricsdocs.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/metrics"
```
Comment (nunnatsa) [inline tools/metricsdocs/metricsdocs.go:6]: goimports here please

**Review by nunnatsa — COMMENTED** [review]
I'm a bit confused by this PR. I don't understand why we should stop the well tested and maintained Prometheus handler and write our own. I don't understand why we need to start a server just to get the already known metrics. I find it over engineering and I would look for much more simple solution. To me, even writing a static file is preferred.

**Comment by machadovilaca** [comment]
@nunnatsa
Since this is part of a larger tool to be later added in the [kubevirt/monitoring](https://github.com/kubevirt/monitoring) repository, it would be good to have an automated way to generate metrics documentation similarly in all KubeVirt components repositories (such as [CDI#2043](https://github.com/kubevirt/containerized-data-importer/pull/2043) and [CNAO#1100](https://github.com/kubevirt/cluster-network-addons-operator/pull/1100)) that the tool could easily execute to ensure all metrics docs exist and are updated.

Also, we follow the same approach as seen in [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt) ([Handler](https://github.com/kubevirt/kubevirt/blob/4b704a0ee1857ab7de0ca73a8c3bdbf0f96b5e12/pkg/monitoring/domainstats/prometheus/prometheus.go#L544), [metrics docs tool](https://github.com/kubevirt/kubevirt/blob/4b704a0ee1857ab7de0ca73a8c3bdbf0f96b5e12/tools/doc-generator/doc-generator.go#L56)), since it provides the easiest way to fetch the metrics information programmatically without major changes in any of the codebases.

**Comment by nunnatsa** [comment]
@machadovilaca - there are 4 code smells in the sonarCloud report: https://sonarcloud.io/project/issues?id=kubevirt_hyperconverged-cluster-operator&pullRequest=1630&resolved=false&types=CODE_SMELL 

Please try to address them.

**Comment by machadovilaca** [comment]
@nunnatsa I think those code smells are unrelated to the changes in this PR. Should I still address them here?

**Comment by nunnatsa** [comment]
These 4 smells are from this PR but are not major. We'll fix them later.

/lgtm
/approve

**Pull request merged**